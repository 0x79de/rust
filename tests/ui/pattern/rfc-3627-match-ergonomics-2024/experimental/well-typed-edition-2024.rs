//@ revisions: stable2021 classic2021 structural2021 classic2024 structural2024
//@[stable2021] edition: 2021
//@[classic2021] edition: 2021
//@[structural2021] edition: 2021
//@[classic2024] edition: 2024
//@[structural2024] edition: 2024
//@[classic2024] run-pass
//@[structural2024] run-pass
//! Test cases for well-typed patterns in edition 2024. These are in their own file to ensure we
//! pass both HIR typeck and MIR borrowck, as we may skip the latter if grouped with failing tests.
#![allow(incomplete_features, unused_mut)]
#![cfg_attr(any(classic2021, classic2024), feature(ref_pat_eat_one_layer_2024))]
#![cfg_attr(any(structural2021, structural2024), feature(ref_pat_eat_one_layer_2024_structural))]

pub fn main() {
    // Tests not using match ergonomics. These should always succeed with the same bindings.
    if let Some(&Some(&mut ref x)) = Some(&Some(&mut 0)) {
        let _: &u32 = x;
    }

    // Tests for differences in how many layers of reference are eaten by reference patterns
    if let Some(Some(&x)) = &Some(Some(&0)) {
        #[cfg(any(stable2021, classic2021, structural2021))] let _: u32 = x;
        #[cfg(any(classic2024, structural2024))] let _: &u32 = x;
    }
    if let Some(&Some(x)) = &mut Some(&Some(0)) {
        // This additionally tests that `&` patterns can eat inherited `&mut` refs.
        // This is possible on stable when the real reference being eaten is of a `&` type.
        #[cfg(any(stable2021, classic2021, structural2021))] let _: u32 = x;
        #[cfg(any(classic2024, structural2024))] let _: &u32 = x;
    }
    if let Some(Some(&&x)) = &Some(Some(&0)) {
        //[stable2021,classic2021,structural2021]~^ mismatched types
        //[stable2021,classic2021,structural2021]~| expected integer, found `&_`
        #[cfg(any(classic2024, structural2024))] let _: u32 = x;
    }

    // Tests for eating a lone inherited reference
    if let Some(Some(&x)) = &Some(&Some(0)) {
        //[stable2021,classic2021,structural2021]~^ mismatched types
        //[stable2021,classic2021,structural2021]~| expected integer, found `&_`
        #[cfg(any(classic2024, structural2024))] let _: u32 = x; // TODO: this should hold on `classic2021` and `structural2021`
    }
    if let Some(&Some(x)) = &Some(Some(0)) {
        //[stable2021,classic2021,structural2021]~^ mismatched types
        //[stable2021,classic2021,structural2021]~| expected `Option<{integer}>`, found `&_`
        #[cfg(any(classic2024, structural2024))] let _: u32 = x; // TODO: this should hold on `classic2021` and `structural2021`
    }
    if let Some(Some(&mut x)) = &mut Some(&mut Some(0)) {
        //[stable2021,classic2021,structural2021]~^ mismatched types
        //[stable2021,classic2021,structural2021]~| expected integer, found `&mut _`
        #[cfg(any(classic2024, structural2024))] let _: u32 = x; // TODO: this should hold on `classic2021` and `structural2021`
    }

    // Tests for `&` patterns matching real `&mut` reference types
    if let Some(&Some(&x)) = Some(&Some(&mut 0)) {
        //[stable2021]~^ mismatched types
        //[stable2021]~| types differ in mutability
        #[cfg(any(classic2021, structural2021, classic2024, structural2024))] let _: u32 = x;
    }

    // Tests for eating only one layer and also eating a lone inherited reference
    if let Some(&Some(&x)) = &Some(&Some(0)) {
        //[stable2021,classic2021,structural2021]~^ mismatched types
        //[stable2021,classic2021,structural2021]~| expected integer, found `&_`
        #[cfg(any(classic2024, structural2024))] let _: u32 = x;
    }

    // Tests for `&` matching a lone inherited possibly-`&mut` reference
    if let Some(&Some(Some(&x))) = &Some(Some(&mut Some(0))) {
        //[stable2021,classic2021,structural2021]~^ mismatched types
        //[stable2021]~| expected `Option<&mut Option<{integer}>>`, found `&_`
        #[cfg(any(classic2024, structural2024))] let _: u32 = x; // TODO: this should hold on `classic2021` and `structural2021`
    }
    if let Some(&Some(x)) = &mut Some(Some(0)) {
        //[stable2021,classic2021,structural2021]~^ mismatched types
        //[stable2021]~| expected `Option<{integer}>`, found `&_`
        #[cfg(any(classic2024, structural2024))] let _: u32 = x; // TODO: this should hold on `classic2021` and `structural2021`
    }

    // Tests eating one layer, eating a lone inherited ref, and `&` eating `&mut` (realness varies)
    if let Some(&Some(&x)) = &Some(&mut Some(0)) {
        //[stable2021,classic2021,structural2021]~^ mismatched types
        //[stable2021]~| types differ in mutability
        //[classic2021,structural2021]~| expected integer, found `&_`
        #[cfg(any(classic2024, structural2024))] let _: u32 = x;
    }
    if let Some(&Some(&x)) = &mut Some(&Some(0)) {
        //[stable2021,classic2021,structural2021]~^ mismatched types
        //[stable2021,classic2021,structural2021]~| expected integer, found `&_`
        #[cfg(any(classic2024, structural2024))] let _: u32 = x;
    }

    // Tests for eat-inner rulesets matching on the outer reference if matching on the inner
    // reference causes a mutability mismatch, i.e. `Deref(EatInner, FallbackToOuter)`:
    let [&mut x] = &mut [&0];
    //[stable2021,classic2021,structural2021]~^ mismatched types
    //[stable2021,classic2021,structural2021]~| types differ in mutability
    // TODO: on `structural2021` `x` should have type `u32`
    #[cfg(any(classic2024, structural2024))] let _: &u32 = x;

    let [&mut ref x] = &mut [&0];
    //[stable2021,classic2021,structural2021]~^ mismatched types
    //[stable2021,classic2021,structural2021]~| types differ in mutability
    // TODO: on `structural2021` `x` should have type `&u32`
    #[cfg(any(classic2024, structural2024))] let _: &&u32 = x;

    let [&mut ref mut x] = &mut [&0];
    //[stable2021,classic2021,structural2021]~^ mismatched types
    //[stable2021,classic2021,structural2021]~| types differ in mutability
    // TODO: this should be a mut borrow behind shared borrow error on `structural2021`
    #[cfg(any(classic2024, structural2024))] let _: &mut &u32 = x;

    let [&mut mut x] = &mut [&0];
    //[stable2021,classic2021,structural2021]~^ mismatched types
    //[stable2021,classic2021,structural2021]~| types differ in mutability
    // TODO: on `structural2021` `x` should have type `u32`
    #[cfg(any(classic2024, structural2024))] let _: &u32 = x;

    let [&mut &x] = &mut [&0];
    //[stable2021,classic2021,structural2021]~^ mismatched types
    //[stable2021,classic2021,structural2021]~| types differ in mutability
    // TODO: [structural2021]~| expected integer, found `&_`
    #[cfg(any(classic2024, structural2024))] let _: u32 = x;

    let [&mut &ref x] = &mut [&0];
    //[stable2021,classic2021,structural2021]~^ mismatched types
    //[stable2021,classic2021,structural2021]~| types differ in mutability
    // TODO: [structural2021]~| expected integer, found `&_`
    #[cfg(any(classic2024, structural2024))] let _: &u32 = x;

    let [&mut &(mut x)] = &mut [&0];
    //[stable2021,classic2021,structural2021]~^ mismatched types
    //[stable2021,classic2021,structural2021]~| types differ in mutability
    // TODO: [structural2021]~| expected integer, found `&_`
    #[cfg(any(classic2024, structural2024))] let _: u32 = x;
}
