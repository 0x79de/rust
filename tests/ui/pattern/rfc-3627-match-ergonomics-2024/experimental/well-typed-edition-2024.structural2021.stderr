error[E0308]: mismatched types
  --> $DIR/well-typed-edition-2024.rs:32:23
   |
LL |     if let Some(Some(&&x)) = &Some(Some(&0)) {
   |                       ^^     --------------- this expression has type `&Option<Option<&{integer}>>`
   |                       |
   |                       expected integer, found `&_`
   |
   = note:   expected type `{integer}`
           found reference `&_`
help: consider removing `&` from the pattern
   |
LL -     if let Some(Some(&&x)) = &Some(Some(&0)) {
LL +     if let Some(Some(&x)) = &Some(Some(&0)) {
   |

error[E0308]: mismatched types
  --> $DIR/well-typed-edition-2024.rs:39:22
   |
LL |     if let Some(Some(&x)) = &Some(&Some(0)) {
   |                      ^^     --------------- this expression has type `&Option<&Option<{integer}>>`
   |                      |
   |                      expected integer, found `&_`
   |
   = note:   expected type `{integer}`
           found reference `&_`
help: consider removing `&` from the pattern
   |
LL -     if let Some(Some(&x)) = &Some(&Some(0)) {
LL +     if let Some(Some(x)) = &Some(&Some(0)) {
   |

error[E0308]: mismatched types
  --> $DIR/well-typed-edition-2024.rs:44:17
   |
LL |     if let Some(&Some(x)) = &Some(Some(0)) {
   |                 ^^^^^^^^    -------------- this expression has type `&Option<Option<{integer}>>`
   |                 |
   |                 expected `Option<{integer}>`, found `&_`
   |
   = note:   expected enum `Option<{integer}>`
           found reference `&_`

error[E0308]: mismatched types
  --> $DIR/well-typed-edition-2024.rs:49:22
   |
LL |     if let Some(Some(&mut x)) = &mut Some(&mut Some(0)) {
   |                      ^^^^^^     ----------------------- this expression has type `&mut Option<&mut Option<{integer}>>`
   |                      |
   |                      expected integer, found `&mut _`
   |
   = note:           expected type `{integer}`
           found mutable reference `&mut _`
note: to declare a mutable binding use: `mut x`
  --> $DIR/well-typed-edition-2024.rs:49:22
   |
LL |     if let Some(Some(&mut x)) = &mut Some(&mut Some(0)) {
   |                      ^^^^^^
help: consider removing `&mut` from the pattern
   |
LL -     if let Some(Some(&mut x)) = &mut Some(&mut Some(0)) {
LL +     if let Some(Some(x)) = &mut Some(&mut Some(0)) {
   |

error[E0308]: mismatched types
  --> $DIR/well-typed-edition-2024.rs:63:23
   |
LL |     if let Some(&Some(&x)) = &Some(&Some(0)) {
   |                       ^^     --------------- this expression has type `&Option<&Option<{integer}>>`
   |                       |
   |                       expected integer, found `&_`
   |
   = note:   expected type `{integer}`
           found reference `&_`
help: consider removing `&` from the pattern
   |
LL -     if let Some(&Some(&x)) = &Some(&Some(0)) {
LL +     if let Some(&Some(x)) = &Some(&Some(0)) {
   |

error[E0308]: mismatched types
  --> $DIR/well-typed-edition-2024.rs:70:17
   |
LL |     if let Some(&Some(Some(&x))) = &Some(Some(&mut Some(0))) {
   |                 ^^^^^^^^^^^^^^^    ------------------------- this expression has type `&Option<Option<&mut Option<{integer}>>>`
   |                 |
   |                 expected `Option<&mut Option<{integer}>>`, found `&_`
   |
   = note:   expected enum `Option<&mut Option<{integer}>>`
           found reference `&_`

error[E0308]: mismatched types
  --> $DIR/well-typed-edition-2024.rs:75:17
   |
LL |     if let Some(&Some(x)) = &mut Some(Some(0)) {
   |                 ^^^^^^^^    ------------------ this expression has type `&mut Option<Option<{integer}>>`
   |                 |
   |                 expected `Option<{integer}>`, found `&_`
   |
   = note:   expected enum `Option<{integer}>`
           found reference `&_`

error[E0308]: mismatched types
  --> $DIR/well-typed-edition-2024.rs:82:23
   |
LL |     if let Some(&Some(&x)) = &Some(&mut Some(0)) {
   |                       ^^     ------------------- this expression has type `&Option<&mut Option<{integer}>>`
   |                       |
   |                       expected integer, found `&_`
   |
   = note:   expected type `{integer}`
           found reference `&_`
help: consider removing `&` from the pattern
   |
LL -     if let Some(&Some(&x)) = &Some(&mut Some(0)) {
LL +     if let Some(&Some(x)) = &Some(&mut Some(0)) {
   |

error[E0308]: mismatched types
  --> $DIR/well-typed-edition-2024.rs:88:23
   |
LL |     if let Some(&Some(&x)) = &mut Some(&Some(0)) {
   |                       ^^     ------------------- this expression has type `&mut Option<&Option<{integer}>>`
   |                       |
   |                       expected integer, found `&_`
   |
   = note:   expected type `{integer}`
           found reference `&_`
help: consider removing `&` from the pattern
   |
LL -     if let Some(&Some(&x)) = &mut Some(&Some(0)) {
LL +     if let Some(&Some(x)) = &mut Some(&Some(0)) {
   |

error[E0308]: mismatched types
  --> $DIR/well-typed-edition-2024.rs:96:10
   |
LL |     let [&mut x] = &mut [&0];
   |          ^^^^^^    --------- this expression has type `&mut [&{integer}; 1]`
   |          |
   |          types differ in mutability
   |
   = note:      expected reference `&{integer}`
           found mutable reference `&mut _`
note: to declare a mutable binding use: `mut x`
  --> $DIR/well-typed-edition-2024.rs:96:10
   |
LL |     let [&mut x] = &mut [&0];
   |          ^^^^^^
help: consider removing `&mut` from the pattern
   |
LL -     let [&mut x] = &mut [&0];
LL +     let [x] = &mut [&0];
   |

error[E0308]: mismatched types
  --> $DIR/well-typed-edition-2024.rs:102:10
   |
LL |     let [&mut ref x] = &mut [&0];
   |          ^^^^^^^^^^    --------- this expression has type `&mut [&{integer}; 1]`
   |          |
   |          types differ in mutability
   |
   = note:      expected reference `&{integer}`
           found mutable reference `&mut _`
note: to declare a mutable binding use: `mut x`
  --> $DIR/well-typed-edition-2024.rs:102:10
   |
LL |     let [&mut ref x] = &mut [&0];
   |          ^^^^^^^^^^
help: consider removing `&mut` from the pattern
   |
LL -     let [&mut ref x] = &mut [&0];
LL +     let [ref x] = &mut [&0];
   |

error[E0308]: mismatched types
  --> $DIR/well-typed-edition-2024.rs:108:10
   |
LL |     let [&mut ref mut x] = &mut [&0];
   |          ^^^^^^^^^^^^^^    --------- this expression has type `&mut [&{integer}; 1]`
   |          |
   |          types differ in mutability
   |
   = note:      expected reference `&{integer}`
           found mutable reference `&mut _`
note: to declare a mutable binding use: `mut x`
  --> $DIR/well-typed-edition-2024.rs:108:10
   |
LL |     let [&mut ref mut x] = &mut [&0];
   |          ^^^^^^^^^^^^^^
help: consider removing `&mut` from the pattern
   |
LL -     let [&mut ref mut x] = &mut [&0];
LL +     let [ref mut x] = &mut [&0];
   |

error[E0308]: mismatched types
  --> $DIR/well-typed-edition-2024.rs:114:10
   |
LL |     let [&mut mut x] = &mut [&0];
   |          ^^^^^^^^^^    --------- this expression has type `&mut [&{integer}; 1]`
   |          |
   |          types differ in mutability
   |
   = note:      expected reference `&{integer}`
           found mutable reference `&mut _`
note: to declare a mutable binding use: `mut x`
  --> $DIR/well-typed-edition-2024.rs:114:10
   |
LL |     let [&mut mut x] = &mut [&0];
   |          ^^^^^^^^^^
help: consider removing `&mut` from the pattern
   |
LL -     let [&mut mut x] = &mut [&0];
LL +     let [mut x] = &mut [&0];
   |

error[E0308]: mismatched types
  --> $DIR/well-typed-edition-2024.rs:120:10
   |
LL |     let [&mut &x] = &mut [&0];
   |          ^^^^^^^    --------- this expression has type `&mut [&{integer}; 1]`
   |          |
   |          types differ in mutability
   |
   = note:      expected reference `&{integer}`
           found mutable reference `&mut _`

error[E0308]: mismatched types
  --> $DIR/well-typed-edition-2024.rs:126:10
   |
LL |     let [&mut &ref x] = &mut [&0];
   |          ^^^^^^^^^^^    --------- this expression has type `&mut [&{integer}; 1]`
   |          |
   |          types differ in mutability
   |
   = note:      expected reference `&{integer}`
           found mutable reference `&mut _`

error[E0308]: mismatched types
  --> $DIR/well-typed-edition-2024.rs:132:10
   |
LL |     let [&mut &(mut x)] = &mut [&0];
   |          ^^^^^^^^^^^^^    --------- this expression has type `&mut [&{integer}; 1]`
   |          |
   |          types differ in mutability
   |
   = note:      expected reference `&{integer}`
           found mutable reference `&mut _`

error: aborting due to 16 previous errors

For more information about this error, try `rustc --explain E0308`.
